### PR TITLE
[PRM-640]: added Upscan connector plus parser

### PR DESCRIPTION
**Pull request by BackRowOla**

### CODE REVIEW

File: app/connectors/httpParsers/UpscanInitiateHttpParser.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    override val body: String = "Invalid JSON received"
+  }
+
+  case object BadRequest extends ErrorResponse {
+    override val status: Int = BAD_REQUEST
+    override val body: String = "incorrect Json body sent"
+  }
+
+  case class UnexpectedFailure(override val status: Int, override val body: String) extends ErrorResponse
+}
```
Comment (LewisMitchell) [inline app/connectors/httpParsers/UpscanInitiateHttpParser.scala:63]: I'm not a massive fan of mixing case object / classes in the same file. If you want to keep it in the same file then you could make ErrorResponse a sealed trait, but you could argue that these objects are quite generic and could be used elsewhere?

**TL:DR; could you make ErrorResponse a sealed trait or move it to a different file?**

File: app/models/upscan/UpscanInitiateResponseModel.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
+
+package models.upscan
+
+import play.api.libs.json.{Format, JsError, JsResult, JsString, JsSuccess, JsValue, Json, Reads}
+
+case class UpscanInitiateResponseModel(
+                                   reference: Reference,
+                                   uploadRequest: UploadFormTemplateRequest
+                                 )
+case class Reference(value: String)
```
Comment (LewisMitchell) [inline app/models/upscan/UpscanInitiateResponseModel.scala:25]: I'm not sure in the need for a wrapper class for the file reference? Up to you if you want to keep it but it's just a String?

File: app/models/upscan/UpscanInitiateResponseModel.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  implicit val referenceFormat: Format[Reference] = new Format[Reference] {
+
+    override def reads(json: JsValue): JsResult[Reference] = json.validate[String] match {
+      case JsSuccess(reference, _) => JsSuccess(Reference(reference))
+      case error: JsError => error
+    }
+
+    override def writes(reference: Reference): JsValue = JsString(reference.value)
+  }
```
Comment (LewisMitchell) [inline app/models/upscan/UpscanInitiateResponseModel.scala:39]: I think String's have implicit JSON formatters anyway, so I don't think this is needed?

File: app/models/upscan/UpscanInitiateResponseModel.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  implicit val jsonFormatUploadForm: Format[UploadFormTemplateRequest] = Json.format[UploadFormTemplateRequest]
+
+  implicit val JsonReadsForModel: Reads[UpscanInitiateResponseModel] = new Reads[UpscanInitiateResponseModel] {
```
Comment (LewisMitchell) [inline app/models/upscan/UpscanInitiateResponseModel.scala:43]: Are we using this `JsonReadsForModel`? I see you have `formats` further up and that `formats` is being used in the connector?

**Review by LewisMitchell — CHANGES_REQUESTED** [review]
Could you add some tests for the connector, the model and the parser? Also could you update app-config-base with the upscan-initiate config? Other than that, looks good.

File: app/connectors/httpParsers/UpscanInitiateHttpParser.scala (hunk 1 above)
Comment (BackRowOla) [reply app/connectors/httpParsers/UpscanInitiateHttpParser.scala:63]: Changed error response to a sealed trait

File: app/models/upscan/UpscanInitiateResponseModel.scala (hunk 2 above)
Comment (BackRowOla) [reply app/models/upscan/UpscanInitiateResponseModel.scala:25]: Makes sense, will change to a string and can remove the format in the object

File: app/models/upscan/UpscanInitiateResponseModel.scala (hunk 3 above)
Comment (BackRowOla) [reply app/models/upscan/UpscanInitiateResponseModel.scala:39]: addressed in comment above

File: app/models/upscan/UpscanInitiateResponseModel.scala (hunk 4 above)
Comment (BackRowOla) [reply app/models/upscan/UpscanInitiateResponseModel.scala:43]: Was meant to use the reads for model rather than the formats 😅

**Review by BackRowOla — COMMENTED** [review]
appConfig-base PR: https://github.com/hmrc/app-config-base/pull/6171

**Review by LewisMitchell — APPROVED** [review]
Looks good, just need to add what I mentioned in Slack and then I'll merge.

**Pull request merged**